File: python/prime_count_src/primenumbers.py
import math

def create_sieve(s, n):
    s.append(False) #0
    s.append(False) #1
    s.append(True)  #2
    for i in range(3, n+1):
        s.append(i%2 != 0)


def compute_sieve(s):
    k = int(math.sqrt(len(s)))
    for i in range(3, k, 2):
        if s[i]:
            j = i + i
            while j < len(s):
                s[j] = False
                j = j + i


def count_primes(s):
    c = 0
    for i in s:
        if i:
            c += 1
    return c

File: c/src/prime_count/primenumbers.c
#include <math.h>
#include "primenumbers.h"

void compute_sieve(bool s[], unsigned n) {
    unsigned i, j, k;
    for (i=3; i<=n; i++) { s[i] = i%2!=0; }
    s[0] = false;
    s[1] = false;
    s[2] = true;
    k = sqrt(n);

    for (i=3; i<=k; i+=2) {
        if (s[i]) {
            for (j=i*i; j<=n; j+=i) {
                s[j] = false;
            }
        }
    }
}

unsigned count_primes(bool s[], unsigned n) {
    int i, c = 0;
    if (n >= 2) {
        c++;
        for (i=3; i<=n; i+=2) {
            if (s[i]) { c++; }
        }
    } 
    return c;
}
